src/common/libflux/message.c

```diff
@@ -444,50 +460,50 @@ static int zmsg_to_msg (flux_msg_t *msg, zmsg_t *zmsg)
     }
     msg->flags = proto_data[PROTO_OFF_FLAGS];
 
-    zf = zmsg_first (zmsg);
     if ((msg->flags & FLUX_MSGFLAG_ROUTE)) {
-        if (!zf) {
+        if (index == iovcnt) {
             errno = EPROTO;
             return -1;
         }
-        while (zf && zframe_size (zf) > 0) {
+        while ((index < iovcnt) && iov[index].size > 0) {
             if (msg_append_route (msg,
-                                  (char *)zframe_data (zf),
-                                  zframe_size (zf)) < 0)
+                                  (char *)iov[index].data,
+                                  iov[index].size) < 0)
                 return -1;
-            zf = zmsg_next (zmsg);
+            index++;
         }
-        if (zf)
-            zf = zmsg_next (zmsg);
+        if (index < iovcnt)
+            index++;
     }
     if ((msg->flags & FLUX_MSGFLAG_TOPIC)) {
-        if (!zf) {
+        if (index == iovcnt) {
             errno = EPROTO;
             return -1;
         }
-        if (!(msg->topic = zframe_strdup (zf))) {
+        if (!(msg->topic = strndup ((char *)iov[index].data,
+                                    iov[index].size))) {
             errno = ENOMEM;
             return -1;
         }
-        if (zf)
-            zf = zmsg_next (zmsg);
+        if (index < iovcnt)
+            index++;
     }
     if ((msg->flags & FLUX_MSGFLAG_PAYLOAD)) {
-        if (!zf) {
+        if (index == iovcnt) {
             errno = EPROTO;
             return -1;
         }
-        msg->payload_size = zframe_size (zf);
+        msg->payload_size = iov[index].size;
         if (!(msg->payload = malloc (msg->payload_size))) {
             errno = ENOMEM;
             return -1;
         }
```

Comment:
`malloc()` already sets errno (I realize you didn't touch this line :-)

It might be good to throw in EPROTO if the payload size is zero, otherwise `malloc()` would return NULL without errno set.   The existing code would at least set errno to _something_ but ENOMEM is the wrong one.

Reply:
perhaps we should just check for payload size > 0 in `flux_msg_set_payload()`, so we know its > 0 in all code later.

Reply:
oh actually, `flux_msg_set_payload()` already ensures payload size > 0, so I think we're safe here.